Improved plugin system + added new plugins and type

dtelyukh opened the pull request:
Two features: 1) turning off attributes` names default prefixes; 2) excluding some attributes from decoder (for example, xmlns:xsi or xsi:noNamespaceSchemaLocation)

I use it so
```
	root := &xj.Node{}
	dec := xj.NewDecoder(strings.NewReader(doc))
	dec.TurnOffDefaultPrefixes()
	dec.ExcludeAttributes([]string{"xsi", "noNamespaceSchemaLocation"})
	dec.Decode(root)

	buf := new(bytes.Buffer)
	e := xj.NewEncoder(buf)
	e.Encode(root)

	fmt.Println(buf.String())
```

For the encoder added ChildrenAlwaysAsArray flag to has output like this

Input 
```
        <Corrections>
          <Correction Text="Not acceptable"/>
        </Corrections>
```

Output
```
"Corrections":{
  "Correction":[
    {
      "Text":"Not acceptable"
    }
  ]
}
```

basgys commented:
Hi @dtelyukh, thanks for your contribution. I haven't had a chance to review your code yet.

However, I can already say this:

1. No, we're not going to turn off the default prefix by default. You're not the first one to ask to change the prefix or turn it off. As I said in other PRs/Issues, the prefix is meant to avoid name clashes between attributes and children. If we turn it off, it can quietly discard data. You are free to decide to not use name prefixes if you are confident that there won't be a conflict, but I won't make it the default option.
2. That's nice!
3. ChildrenAlwaysAsArray. I can understand the idea behind this change, but the design is pretty bad in my opinion. Turning all children into arrays is way too limited. I'd rather go for an option that allow to flag attributes that are arrays. This could be couple with your option 2.

Here is how it could work:

```go
  xml := strings.NewReader(`...`)
  json, err := xj.Convert(
    xml, 
    xj.WithAttrPrefix(""),
    xj.WithNodes(
      xj.Node("//book//authors", xj.ElementType(xj.Array)), // Element
      xj.Node("//book/@id", xj.Exclude()), // @Attribute      
    ),
  )
```

What do you think?

dtelyukh commented:
Thank you for such detailed answer very much. :+1: 

1. Sorry for misleading you. I don't want  to turn off the default prefix by default. I wrote TurnOffDefaultPrefixes function, which can turn it off optionally. Look at the PR, please.

2. --

3. Here is also misunderstanding. ChildrenAlwaysAsArray is only for specific node, not for the whole decoder.

But looking at all I should write plugins to apply mentioned flags following your design.

So I found that `WithAttrPrefix` and `WithContentPrefix` are buggy, cause it should be
```
func (a *attrPrefixer) AddTo(e *Encoder) *Encoder {
	e.attributePrefix = string((*a))
	return e
}
```
instead of
```
func (a *attrPrefixer) AddTo(e *Encoder) {
	e.attributePrefix = string((*a))
}
```
Also anyway `xj.WithAttrPrefix("")` has no effect cause it applies to the encoder, but it should be applied to the decoder. The same is true for `Exclude()` and `ChildrenAlwaysAsArray`.
So we need apply plugins to the decoder too. `Convert` function could be looked like this
```
func Convert(r io.Reader, dps []decoderPlugin, eps []encoderPlugin) (*bytes.Buffer, error) {
…
}
```

What do you think?

basgys commented:
1. Ok good. `TurnOffDefaultPrefixes` is redundant then since you have `WithAttrPrefix`
2. -
3. - 

**So I found that WithAttrPrefix and WithContentPrefix are buggy**

Can you explain me how your change fix the faulty behaviour? Also, can you tell me more about the faulty behaviour?

**has no effect cause it applies to the encoder, but it should be applied to the decoder.**

Yes indeed. We should have a plugin system that can be applied to the encoding, decoding, or both. I would suggest to write a `plugin` interface that applies to both parts then to avoid the double parameters (one for decoding and one for encoding).

Sorry for the late answer, I don't have a lot of time to spend on this side project at the moment. Do you have an idea how to improve this plugin system or shall I draft something?

dtelyukh commented:
Hi @basgys

I changed the PR:

- improved plugin system that can be applied to the encoding, decoding, or both
- added new plugins `xj.ExcludeAttributes`, `xj.WithNodes`
- fixed `xj.WithAttrPrefix` and `xj.WithContentPrefix` plugins
- added ability to use empty prefixes
- added new type `xj.Null`
- added more unit-tests

My example:
```
  xml := strings.NewReader(`...`)
  json, err := xj.Convert(
    xml, 
    xj.WithAttrPrefix(""),
    xj.ExcludeAttributes([]string{"xsi", "noNamespaceSchemaLocation"}),
    xj.WithNodes(
	xj.NodePlugin("book.authors", xj.ToArray()),
    ),
  )
```

basgys commented:
@dtelyukh Excellent! I quickly reviewed your changes. I will spend a bit more time reviewing it later this week before merging the PR.

basgys commented:
@dtelyukh Sorry for the delay. Thanks for your work 🥇
